Changelog 163 (#143)
intellij-rust/intellij-rust.github.io: _posts/2022-01-17-changelog-163.markdown

---
layout: post
title: "IntelliJ Rust Changelog #163"
date: 2022-01-17 13:00:00 +0300
---

## New Features

* [#8018] Add a widget for the [external linter](https://plugins.jetbrains.com/plugin/8182-rust/docs/rust-code-analysis.html#call-external-linter).
Widget indicates whether the linter is turned on. Also, the widget shows a notification when linter significantly affects the code analysis performance
//TODO: gif

* [#8132] Add `Extract trait` refactoring (`Refactor | Extract trait`)
//TODO: gif

* [#8315] Complete the second path segment if the first segment is unresolved but can be imported.
//TODO: gif

> * [#8284] Provide completion for out-of-scope items when invoked at an empty place

Reviewer: "The completion can be invoked by Ctrl + Space"

* [#8119] Enable new name resolution inside functions bodies
// Feature or fix?

## Fixes

* [#8362] Do not produce a false-positive error about missing format macro arguments without `format_args_capture` feature starting with 1.58

* [#8334] Don't perform cargo refresh when another IDE project become trusted

* [#8318] Expand proc macros from `dev-dependencies`. Note, procedural macro expansion is still under development, and it's disabled by default for now. To turn it on, enable `org.rust.cargo.evaluate.build.scripts` and `org.rust.macros.proc` [experimental features](https://plugins.jetbrains.com/plugin/8182-rust/docs/rust-faq.html#experimental-features)

* [#8297] Fix default parameters handling in `Wrong number of type arguments` inspection

* [#8291] Improve sorting of auto-import variants
// Feature?

* [#8288] Fix auto-import from `alloc` crate when using [`#![no_std]`](https://docs.rust-embedded.org/book/intro/no-std.html)

* [#8286] Don't show auto-import from auxiliary stdlib crates such as `unwind`

* [#8279] Fix resolve of imports inside detached files

> * [#8269] Correctly parenthesize block-like expressions (e.g. `match`) in `Inline Value` refactoring (by [@Kobzol])

Reviewer 2: Link and shortcut?

* [#8268] Fix type inference if there are dependencies named the same as stdlib crates

> * [#8263] Infer generic parameter types in Convert closure to function intention (by [@Kobzol])

Reviewer 2: ```suggestion
* [#8263] Infer generic parameter types in `Convert closure to function` intention (by [@Kobzol])
```

> * [#8230] "Introduce local variable" no longer moves the new binding into closure bodies if it is used outside of them (by [@afetisov])

Reviewer 2: ```suggestion
* [#8230] `Introduce local variable` refactoring no longer moves the new binding into closure bodies if it is used outside of them (by [@afetisov])
```

Reviewer 2: Maybe also add a link to our doc: https://plugins.jetbrains.com/plugin/8182-rust/docs/rust-refactorings.html#extractvar-refactoring ?

> * [#8228] Do not offer the `Add remaining patterns fix` when pattern matching on an enum marked with `#[non_exhaustive]` if there are no wildcard arms in the `match` expression (by [@Kobzol])

Reviewer 2: ```suggestion
* [#8228] Do not offer the `Add remaining patterns` quick-fix when pattern matching on an enum marked with `#[non_exhaustive]` if there are no wildcard arms in the `match` expression (by [@Kobzol])
```

> * [#8216] Support usages in `serde` string attributes like [`#[serde(default = "path")]`](https://serde.rs/field-attrs.html#default--path) in unused imports inspection

Reviewer 2: ```suggestion
* [#8216] Support usages in `serde` string attributes like [`#[serde(default = "path")]`](https://serde.rs/field-attrs.html#default--path) in `Unused import` inspection
```

* [#8187] Support indirect self-references in [`Move` refactoring](https://plugins.jetbrains.com/plugin/8182-rust/docs/rust-refactorings.html#move-refactoring) (`Refactor | Move` or <kbd>F6</kbd>)

* [#8185] Fix handling comments in `Invert if condition` intention

* [#8165] Fix `Invert if condition` intention when there is `return`/`continue` without semicolon

> * [#8133] Fix renaming mod declaration by "Module naming convention" inspection

Reviewer 2: ```suggestion
* [#8133] Fix rename quick-fix for module declaration by `Module naming convention` inspection
```

---
layout: post
title: "IntelliJ Rust Changelog #163"
date: 2022-01-17 13:00:00 +0300
---

## New Features

* [#8018] Add a widget for the [external linter](https://plugins.jetbrains.com/plugin/8182-rust/docs/rust-code-analysis.html#call-external-linter).
Widget indicates whether the linter is turned on. Also, the widget shows a notification when linter significantly affects the code analysis performance.
The external linter settings were moved to `Preferences | Languages & Frameworks | Rust | External Linters`.

> If necessary, the maximum duration for the external linter can be changed via `Registry... | org.rust.external.linter.max.duration` option

Reviewer 2: Let's not mention `Registry` here if possible since it's more about workaround for unexpected cases than real setting. If we decide that users need this setting, we will move it into user-visible panel

>   <img src="/assets/posts/changelog-163/widget_init.png" width="700px"/>
>   <img src="/assets/posts/changelog-163/widget_on.png" width="700px"/>
>   <img src="/assets/posts/changelog-163/widget_disable.png" width="700px"/>

Reviewer 2: ```suggestion
  <img src="/assets/posts/changelog-163/widget_init.png" width="700px"/>
  
  
  <img src="/assets/posts/changelog-163/widget_on.png" width="700px"/>
  
  
  <img src="/assets/posts/changelog-163/widget_disable.png" width="700px"/>
```

otherwise, it's not clear that there are three separate screenshots

* [#6353] Detect `Const cannot refer to static` [E0013](https://doc.rust-lang.org/error-index.html#E0013) compiler error  (by [@zeroeightysix])

* [#8144] Detect [E0747](https://doc.rust-lang.org/error-index.html#E0747) compiler error

* [#8289] Prioritize expressions that can be implicitly coerced to the expected type when suggesting completion variants

* [#8335] Do not run non-syntax inspections on files disabled by `cfg` attributes

* [#8323] Process available items generated by build script and expand available procedural macro calls even if the whole project host code (build scripts and procedural macro libraries) cannot be successfully compiled and executed.
Note, build script evaluation is disabled by default for now. To turn it on, enable `org.rust.cargo.evaluate.build.scripts` [experimental](https://plugins.jetbrains.com/plugin/8182-rust/docs/rust-faq.html#experimental-features) feature

* [#8119] Enable new name resolution inside functions bodies. It allows resolving [macros 2.0](https://rust-lang.github.io/rfcs/1584-macros.html) declared inside function bodies

## Fixes

> * Auto-import fixes:

Reviewer 2: ```suggestion
* Auto-import improvements:
```

* [#8334] Don't perform [project model reloading](https://plugins.jetbrains.com/plugin/8182-rust/docs/rust-cargo-tool-window.html#cargo-load) when another IDE project become [trusted](https://plugins.jetbrains.com/plugin/8182-rust/docs/rust-cargo-tool-window.html#trust-project)

* [#8318] Expand proc macros from `dev-dependencies`. Note, procedural macro expansion is still under development, and it's disabled by default for now. To turn it on, enable `org.rust.cargo.evaluate.build.scripts` and `org.rust.macros.proc` [experimental features](https://plugins.jetbrains.com/plugin/8182-rust/docs/rust-faq.html#experimental-features)

* [#8297] Fix default parameters handling in `Wrong number of type arguments` inspection

* [#8279] Fix resolve of imports inside detached files

* [#8269] Correctly parenthesize block-like expressions (e.g. `match`) in [`Inline Variable` refactoring](https://plugins.jetbrains.com/plugin/8182-rust/docs/rust-refactorings.html#extractvar-refactoring) (by [@Kobzol])

* [#8268] Fix type inference if there are dependencies named the same as stdlib crates

* [#8263] Infer generic parameter types in `Convert closure to function` intention (by [@Kobzol])

> * [#8230] [`Introduce local variable` refactoring](https://plugins.jetbrains.com/plugin/8182-rust/docs/rust-refactorings.html#extractvar-refactoring) no longer moves the new binding into closure bodies if it is used outside of them (by [@afetisov])

Reviewer 2: Note, currently, it's rendered in the following way
![image](https://user-images.githubusercontent.com/2539310/149757326-5cb7b1d7-54f0-4eca-ab79-d32151264518.png)

I can't come up with a better solution than
```suggestion
* [#8230]&nbsp;[`Introduce local variable` refactoring](https://plugins.jetbrains.com/plugin/8182-rust/docs/rust-refactorings.html#extractvar-refactoring) no longer moves the new binding into closure bodies if it is used outside of them (by [@afetisov])
```
